src/signet.cpp

```diff
@@ -0,0 +1,118 @@
+// Copyright (c) 2019-2020 The Bitcoin Core developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#include <signet.h>
+
+#include <consensus/merkle.h>
+#include <consensus/params.h>
+#include <consensus/validation.h>
+#include <hash.h>
+#include <primitives/block.h>
+#include <primitives/transaction.h>
+#include <script/interpreter.h>
+#include <script/standard.h>
+#include <util/system.h>
+
+// Signet block solution checker
+bool CheckBlockSolution(const CBlock& block, const Consensus::Params& consensusParams)
+{
+    std::vector<uint8_t> signet_data;
+    if (!GetWitnessCommitmentSection(block, SIGNET_HEADER, signet_data)) {
+        return error("CheckBlockSolution: Errors in block (block solution missing)");
+    }
+    SimpleSignatureChecker bsc(GetSignetHash(block));
+    CScript challenge(consensusParams.signet_challenge.begin(), consensusParams.signet_challenge.end());
+    CScript solution = CScript(signet_data.begin(), signet_data.end());
+
+    if (!VerifyScript(solution, challenge, nullptr, MANDATORY_SCRIPT_VERIFY_FLAGS, bsc)) {
```

Comment:
People can already kind of malleate blocks by changing the nonce, since it isn't committed to in the signet hash, so it feels like going for the same level of restrictions as in the block checks is sensible, I think. Thoughts @ajtowns ?

Reply:
Since you can already malleate the nonce to get a different block, and can't build on the block without being able to do a valid signature, I guess malleation isn't really a concern, but I think extra flags would make sense. I wonder if exporting  `validation.cpp:GetBlockScriptFlags` and calling it with the genesis block (or making it accept pindexPrev and calling it with nullptr) to get the script flags would make sense.